### PR TITLE
First merge of Migraphx worker 

### PR DESCRIPTION
**Pull request by bpickrel**
# New worker for inference server; Dockerfile updated to install ROCm and Migraphx libraries in support, 2 example Python scripts and automated test case.

*  Migraphx is AMD's library for inference on a GPU.  This adds GPU support to the Inference Server.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -779,6 +781,63 @@ RUN apt-get update \
...
+
+# Install migraphx which supports GPU targets.  At the time of writing this,
+# it was necessary to build migraphx from source but it should be possible to install from repo
+# instead.
+#
+# Add rocm repository
+RUN sh -c 'echo deb [arch=amd64 trusted=yes] http://repo.radeon.com/rocm/apt/5.0/ ubuntu main > /etc/apt/sources.list.d/rocm.list'
+# Install dependencies for migraphx
+RUN apt-get update &&\
+    apt-get install -y sudo git bash build-essential rocm-dev libpython3.6-dev python3-dev python3-pip miopen-hip-dev \
```
Comment (varunsh-xilinx) [inline Dockerfile:797]: build-essential, git, sudo and some other packages are already present. You don't need to install them again. That also allows you to not need to reset gcc below. Once you combine these into one RUN command, you should also delete any packages that are only needed at build-time

File: Dockerfile [hunk 2]
```diff
@@ -779,6 +781,63 @@ RUN apt-get update \
...
+# Workaround broken rocm packages
+RUN ln -s /opt/rocm-* /opt/rocm
+RUN echo "/opt/rocm/lib" > /etc/ld.so.conf.d/rocm.conf
+RUN echo "/opt/rocm/llvm/lib" > /etc/ld.so.conf.d/rocm-llvm.conf
+RUN ldconfig
+# ENV PYTHONPATH=/opt/rocm/lib   # This addition may be needed to import migraphx in Python scripts
+#        for Release version, since it's set up differently than dev
+#
+# Install rbuild
+RUN pip3 install https://github.com/RadeonOpenCompute/rbuild/archive/master.tar.gz
```
Comment (varunsh-xilinx) [inline Dockerfile:808]: Use a commit tag rather than master directly

File: Dockerfile [hunk 3]
```diff
@@ -779,6 +781,63 @@ RUN apt-get update \
...
+# Install MIGraphX from source
+RUN mkdir -p /migraphx
+# RUN cd /migraphx && git clone --depth=1 --branch rocm-5.1.1 https://github.com/ROCmSoftwarePlatform/AMDMIGraphX src
+# at time of writing, a needed bug fix is in the develop branch of AMDMIGraphX
+RUN cd /migraphx && git clone --branch develop https://github.com/ROCmSoftwarePlatform/AMDMIGraphX src
+RUN cd /migraphx/src  && git checkout cb18b0b5722373c49f5c257380af206e13344735
+RUN apt install nano
```
Comment (varunsh-xilinx) [inline Dockerfile:815]: why is nano needed?

File: README.rst [hunk 4]
```diff
@@ -13,10 +13,8 @@
     See the License for the specific language governing permissions and
     limitations under the License.
 
-AMD/Xilinx Inference Server
-===========================
+The AMD/Xilinx Inference Server is an easy-to-use inferencing solution designed for AMD CPUs, GPUs, FPGAs and :github:`Vitis AI <Xilinx/Vitis-AI>`..
```
Comment (varunsh-xilinx) [inline README.rst]: Undo these changes

File: examples/python/custom_processing.py [hunk 5]
```diff
@@ -102,9 +102,16 @@ def main():
         proteus.clients.startHttpServer(8998)
 
     # +load worker:
+<<<<<<< HEAD
```
Comment (varunsh-xilinx) [inline examples/python/custom_processing.py]: Delete these merge conflict artifacts

File: examples/python/hello_migraphx.py [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+            for d in tensor_type.shape.dim:
+                # the dimension may have a definite (integer) value or a symbolic identifier or neither:
+                if (d.HasField("dim_value")):
+                    shape.append(d.dim_value)
+        else:
+            print ("unknown rank", end="")
+        print()
+        break
+
+print('This model\'s shape of input image is ', shape)
```
Comment (varunsh-xilinx) [inline examples/python/hello_migraphx.py:137]: Run `black .` in the repository root to format Python code

File: examples/python/hello_migraphx.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+# it finds a *.onnx file it will compile it and save the compiled model as *.mxr for
+# future use.  It will read
+# the array dimensions and data type from the model.
+
+parameters = proteus.RequestParameters()
+parameters.put("model", modelname)
+# this call requests the server to either find a running instance of the named
+# worker type, or else create one and initialize it with the parameters.  It's
+# currently not possible to search for a worker by specifying the model you want.
+worker_name = client.modelLoad("Migraphx", parameters)
```
Comment (varunsh-xilinx) [inline examples/python/hello_migraphx.py:173]: rename to `workerLoad` after you rebase

File: examples/python/hello_migraphx.py [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+
+# wait for the worker to load and compile model
+ready = False
```
Comment (varunsh-xilinx) [inline examples/python/hello_migraphx.py:176]: this is also simplified on the latest:
```
while not ready:
        ready = client.modelReady(worker_name)
```

File: proteus [hunk 9]
```diff
@@ -638,6 +641,11 @@ class Dockerize:
         if args.dev_base:
             build_args += f"--build-arg DEV_BASE_IMAGE={args.dev_base} "
 
+        if args.migraphx_path:
+            print('migraphx path is ', {args.migraphx_path})
```
Comment (varunsh-xilinx) [inline proteus:646]: delete

File: src/python/src/proteus/rest.py [hunk 10]
```diff
@@ -0,0 +1,349 @@
+# Copyright 2021 Xilinx Inc.
```
Comment (varunsh-xilinx) [inline src/python/src/proteus/rest.py:1]: Delete this file

File: src/proteus/workers/migraphx.cpp [hunk 11]
```diff
@@ -0,0 +1,496 @@
...
+          PROTEUS_IF_LOGGING(logger.info("beginning migraphx eval"));
+          migraphx::api::arguments  migraphx_output = this->prog_.eval(params);
+          PROTEUS_IF_LOGGING(logger.info("finishing migraphx eval"));
+
+          //
+          // Transfer the migraphx results to output
+          //
+          size_t result_size = migraphx_output.size();  // should be 1 item with 1000 (for resnet) categories
+          if(result_size != 1)
+            throw std::length_error("result count from migraphx call was not 1");
```
Comment (varunsh-xilinx) [inline src/proteus/workers/migraphx.cpp:402]: Don't throw in the `doRun()`

**Review by varunsh-xilinx — CHANGES_REQUESTED** [review]
I added an initial review. Thanks for your contribution!

File: src/proteus/workers/migraphx.cpp [hunk 12]
```diff
@@ -0,0 +1,496 @@
...
+  for (const auto& [k, v] : parameters->data()){
+      std::cout << k << " :- ";
+      std::visit([](const auto& x){ std::cout << x; }, v);
+      std::cout << '\n';
+  }
+  if (parameters->has("model")){
+      input_file_ = parameters->get<std::string>("model");
+  }
+  else{
+      PROTEUS_IF_LOGGING(logger.error(
```
[the remark this answers is not in the document]
Comment (bpickrel) [reply src/proteus/workers/migraphx.cpp]: fixed e8e8bdd

File: examples/python/hello_migraphx.py (hunk 6 above)
Comment (bpickrel) [reply examples/python/hello_migraphx.py]: fixed

File: examples/python/hello_migraphx.py (hunk 7 above)
Comment (bpickrel) [reply examples/python/hello_migraphx.py]: fixed e8e8bdd

File: src/proteus/workers/migraphx.cpp [hunk 13]
```diff
@@ -0,0 +1,496 @@
...
+    this->returnBuffers(std::move(batch->input_buffers),
+                        std::move(batch->output_buffers));
+    PROTEUS_IF_LOGGING(logger.debug( "Returned buffers"));
+  }
+  std::cout << "exiting MIGraphXWorker::doInit\n";
+
+  PROTEUS_IF_LOGGING(logger.info("Migraphx ending"));
+}
+
+void MIGraphXWorker::doRelease() {    std::cout << "MIGraphXWorker::doRelease\n";
```
[the remark this answers is not in the document]
Comment (bpickrel) [reply src/proteus/workers/migraphx.cpp]: done  49679bc

File: tests/workers/test_migraphx.cpp [hunk 14]
```diff
@@ -0,0 +1,57 @@
+// Copyright 2022 AMD Inc.
```
[the remark this answers is not in the document]
Comment (bpickrel) [reply tests/workers/test_migraphx.cpp:1]: I get a build error if the file isn't there:
`CMake Error at tests/workers/CMakeLists.txt:32 (add_executable):
  Cannot find source file:
    test_migraphx.cpp
`

File: CMakeLists.txt [hunk 15]
```diff
@@ -124,7 +127,9 @@ add_option("ENABLE_IPO" "Enable interprocedural optimizations" OFF)
 add_option("ENABLE_LINTING" "Enable build-time linting" OFF)
 add_option("ENABLE_TFZENDNN" "Enable TF+ZenDNN worker" ${tfzendnn_FOUND})
 add_option("ENABLE_PTZENDNN" "Enable PT+ZenDNN worker" ${ptzendnn_FOUND})
+add_option("ENABLE_MIGRAPHX" "Enable migraphx worker" ON)
 
+message(STATUS "  PROTEUS_ENABLE_MIGRAPHX: " ${PROTEUS_ENABLE_MIGRAPHX})
```
[the remark this answers is not in the document]
Comment (bpickrel) [reply CMakeLists.txt]: done, commit a097b42

File: examples/python/migraph_resnet.py [hunk 16]
```diff
@@ -0,0 +1,272 @@
...
+    default= os.path.join(root,  r"external/artifacts/migraphx/resnet50-v1-7/resnet50-v1-7.onnx"),
+    help="Location of model file on server",
+)
+parser.add_argument(
+    "--validation_dir",
+    "-v",
+    type=str,
+    required=False,
+    default= os.path.join(root,  r"external/artifacts/migraphx/ILSVRC2012_img_val"),
+    help="The directory containing validation images.  See Mike Vermeulen or Brian Pickrell for image set",
```
[the remark this answers is not in the document]
Comment (bpickrel) [reply examples/python/migraph_resnet.py]: done, commit a097b42

File: proteus (hunk 9 above)
Comment (bpickrel) [reply proteus:645]: done, commit a097b42

File: src/proteus/core/worker_info.cpp [hunk 17]
```diff
@@ -245,6 +245,7 @@ bool WorkerInfo::inputSizeValid(size_t size) const {
 void WorkerInfo::allocate(size_t request_size) {
   // TODO(varunsh): can result in deadlock in manager if allocated num !=
   // request
+
```
[the remark this answers is not in the document]
Comment (bpickrel) [reply src/proteus/core/worker_info.cpp]: done, commit a097b42

File: tests/performance/batching/perf_soft.cpp [hunk 18]
```diff
@@ -139,6 +140,7 @@ class PerfSoftBatcherFixture
 // @pytest.mark.perf(group="batcher")
 TEST_P(PerfSoftBatcherFixture, BasicBatching) {
   const auto [batch_size, num_buffers, delay] = GetParam();
+  (void)num_buffers;  // suppress unused variable warning
```
[the remark this answers is not in the document]
Comment (bpickrel) [reply tests/performance/batching/perf_soft.cpp]: done, commit a097b42

File: tests/workers/CMakeLists.txt [hunk 19]
```diff
@@ -40,7 +44,7 @@ endif()
 
 if(${PROTEUS_ENABLE_AKS})
     target_link_libraries(test_facedetect PRIVATE
-        opencv_core opencv_imgcodecs
+        opencv_core opencv_imgcodecs stdc++fs
```
[the remark this answers is not in the document]
Comment (bpickrel) [reply tests/workers/CMakeLists.txt]: done, commit a097b42

File: Dockerfile [hunk 20]
```diff
@@ -779,6 +781,63 @@ RUN apt-get update \
 
 FROM proteus_install_ptzendnn_${ENABLE_PTZENDNN} as proteus_dev_final
 
+
```
[the remark this answers is not in the document]
Comment (bpickrel) [reply Dockerfile:788]: see proteus line 646 etc.  also, run command lines into one with && operator

[the remark this answers is not in the document]
Comment (bpickrel) [reply Dockerfile:786]: Installation has been combined into fewer RUN stmts with commit 765a0815.  I didn't try to do everything in a single line, for readability (you can't put comments in a long multi-line statement).  ENABLE_MIGRAPHX is used to make migraphx build conditional with commit 567a429ffe

File: Dockerfile (hunk 1 above)
Comment (bpickrel) [reply Dockerfile]: All except the gcc version reset, done. gcc version still needed for builds to work within the Docker container; not clear what's changing it.  Commit 765a081

File: Dockerfile (hunk 20 above)
[the remark this answers is not in the document]
Comment (bpickrel) [reply Dockerfile:786]: Done commit 71b0aa1627

File: Dockerfile (hunk 1 above)
Comment (bpickrel) [reply Dockerfile]: See commit 71b0aa1627.  The gcc version reset turns out to still be necessary.

File: Dockerfile (hunk 2 above)
Comment (bpickrel) [reply Dockerfile]: Cone commit 71b0aa

File: Dockerfile (hunk 3 above)
Comment (bpickrel) [reply Dockerfile]: debug only

**Comment by bpickrel** [comment]
Need to add the file imagenet-simple-labels.json to the tests directory, since QA wants to use it.

**Comment by varunsh-xilinx** [comment]
> Need to add the file imagenet-simple-labels.json to the tests directory, since QA wants to use it.

ZenDNN added the labels [here](https://github.com/Xilinx/inference-server/blob/main/examples/python/utils/imagenet_classes.txt) already

File: proteus [hunk 21]
```diff
@@ -148,7 +148,13 @@ class WriteComposeFile:
         files = glob.glob("/dev/xclmgmt*")
         for file in files:
             devices.append(f"{file}:{file}")
-
+        # migraphx devices, if found
+        kfd = glob.glob('/dev/kfd')
```
Comment (varunsh-xilinx) [inline proteus:152]: Do you know if you need all the files in these directories or some known subset? Just wondering if it's safe to add everything in these directories. I added the other device files only if their name matched some format e.g. render*

File: proteus [hunk 22]
```diff
@@ -148,7 +148,13 @@ class WriteComposeFile:
...
+        dri = glob.glob('/dev/dri')
+        if len(kfd) == 1:
+            devices.append(kfd)
+        # if len(dri) == 1:
+        devices.append(dri)
```
Comment (varunsh-xilinx) [inline proteus:157]: glob returns an array. You should be iterating through the array and adding its contents like the existing devices

File: proteus [hunk 23]
```diff
@@ -1267,7 +1287,7 @@ class Start:
             command += "gdb --args "
         command += f"{args.dir}/{old_build_config}/src/proteus/proteus-server "
         command += " ".join(unknown_args)
-
+
```
Comment (varunsh-xilinx) [inline proteus:1290]: delete

File: src/proteus/bindings/python/setup.py [hunk 24]
```diff
@@ -28,10 +28,7 @@
     version=version,
     license="Apache 2.0",
     packages=setuptools.find_packages("src"),
-    install_requires=[
-        "numpy",
-        "opencv-python-headless",
-    ],
+    install_requires=["numpy", "opencv-python-headless"],
```
Comment (varunsh-xilinx) [inline src/proteus/bindings/python/setup.py:31]: undo

File: src/proteus/core/worker_info.cpp [hunk 25]
```diff
@@ -260,8 +260,6 @@ bool WorkerInfo::inputSizeValid(size_t size) const {
 }
 
 void WorkerInfo::allocate(size_t request_size) {
-  // TODO(varunsh): can result in deadlock in manager if allocated num !=
-  // request
```
Comment (varunsh-xilinx) [inline src/proteus/core/worker_info.cpp:264]: undo

File: src/proteus/workers/CMakeLists.txt [hunk 26]
```diff
@@ -18,6 +18,7 @@ list(APPEND workers
     Echo
     InvertImage
     InvertVideo
+    # Migraphx
```
Comment (varunsh-xilinx) [inline src/proteus/workers/CMakeLists.txt:21]: delete

File: src/proteus/workers/CMakeLists.txt [hunk 27]
```diff
@@ -33,8 +34,14 @@ if(${PROTEUS_ENABLE_PTZENDNN})
     set(workers ${workers} PtZendnn)
 endif()
 
+if(${PROTEUS_ENABLE_MIGRAPHX})
+    set(workers ${workers} Migraphx)
+endif()
+
 foreach( worker ${workers})
-    # convert name to file name: separate by capital letters, add underscores and lowercase
+    set(worker_name worker${worker})
```
Comment (varunsh-xilinx) [inline src/proteus/workers/CMakeLists.txt:42]: undo?

File: tests/performance/batching/perf_soft.cpp [hunk 28]
```diff
@@ -63,6 +63,7 @@ class PerfSoftBatcherFixture
 
   int dequeue() {
     const auto [batch_size, num_buffers, delay] = GetParam();
+    (void)num_buffers;  // suppress unused variable warning
```
Comment (varunsh-xilinx) [inline tests/performance/batching/perf_soft.cpp:66]: undo

File: tests/workers/test_ptzendnn.py [hunk 29]
```diff
@@ -91,10 +91,7 @@ def test_ptzendnn_0(self):
         images = []
         for _ in range(batch):
             images.append(
-                preprocess_pt(
-                    image_path,
-                    input_size=preprocessing["input_size"],
-                )
+                preprocess_pt(image_path, input_size=preprocessing["input_size"])
```
Comment (varunsh-xilinx) [inline tests/workers/test_ptzendnn.py:94]: undo

**Review by varunsh-xilinx — CHANGES_REQUESTED** [review]
Just undo the changes to the other files that are outside this PR. Other changes can be made another time

File: examples/python/hello_migraphx.py [hunk 30]
```diff
@@ -0,0 +1,276 @@
...
+    help="Location of model file on server",
+)
+
+parser.add_argument(
+    "--labels",
+    "-l",
+    type=str,
+    required=False,
+    default=os.path.join(
+        root, r"tests/assets/imagenet_simple_labels.json"
```
[the remark this answers is not in the document]
Comment (bpickrel) [reply examples/python/hello_migraphx.py:88]: I added the file.  I'm leaving it because my script expects the json format.

File: proteus (hunk 21 above)
Comment (bpickrel) [reply proteus:152]: globbing a single file gives a list with 1 member if it exists.  But this code creates a syntax bug and I'm pushing a fix
